plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSegmentPruner.java

```diff
@@ -0,0 +1,193 @@
+                // Not a spooled segment
+                if (handle.isEmpty()) {
+                    continue;
+                }
+
+                if (handle.get().isBefore(expiredBefore)) {
+                    expiredSegments.add(file.location());
+                    if (expiredSegments.size() >= batchSize) {
+                        pruneExpiredSegments(expiredBefore, expiredSegments);
+                    }
+                }
+                else if (filesAreOrdered) {
+                    // First non expired segment was found, no need to check the rest
+                    // since we know that files are lexicographically ordered.
+                    pruneExpiredSegments(expiredBefore, expiredSegments);
+                    return;
+                }
+            }
+            pruneExpiredSegments(expiredBefore, expiredSegments);
+        }
+        catch (IOException e) {
+            log.error(e, "Failed to prune segments");
+        }
+    }
+
+    private void pruneExpiredSegments(Instant expiredBefore, List<Location> expiredSegments)
+    {
+        if (expiredSegments.isEmpty()) {
+            return;
+        }
+
+        try {
+            int batchSize = expiredSegments.size();
+
+            Instant oldest = getExpiration(expiredSegments.getFirst());
+            Instant newest = getExpiration(expiredSegments.getLast());
+            fileSystem.deleteFiles(expiredSegments);
+            log.info("Pruned %d segments expired before %s [oldest: %s, newest: %s]", batchSize, expiredBefore, oldest, newest);
+        }
+        catch (IOException e) {
+            log.warn(e, "Failed to delete %d expired segments", expiredSegments.size());
+        }
+        catch (Exception e) {
+            log.error(e, "Unexpected error while pruning expired segments");
+        }
+    }
+
+    private Optional<Instant> parseHandle(FileEntry file)
+    {
+        try {
+            return Optional.of(getExpiration(file.location()));
+        }
+        catch (Exception e) {
+            return Optional.empty();
+        }
+    }
+
+    private Instant getExpiration(Location location)
+    {
+        return fromStorageObjectName(location.fileName()).expirationTime();
+    }
+
+    private FileIterator orderDetectingIterator(FileIterator delegate)
+    {
+        return new FileIterator()
+        {
+            private FileEntry last;
+
+            @Override
+            public boolean hasNext()
+                    throws IOException
+            {
+                return delegate.hasNext();
+            }
+
+            @Override
+            public FileEntry next()
+                    throws IOException
+            {
+                FileEntry next = delegate.next();
+                // Switch the filesAreOrdered if the files listing order is broken
+                if (!filesAreOrdered && last != null && last.location().fileName().compareTo(next.location().fileName()) > 0) {
```

Comment:
@losipiuk

Reply:
You meant `filesAreOrdered` (not `!filesAreOrdered`) here I assume :P

Reply:
It is best effort. Would rather get rid of violators instead of having logic which is very hard to reason about in case something breaks.

Reply:
oh yeah :)

Reply:
Yeah I know but once the non-sorted order is detected, this class switches to best effort. It's unlikely case for the supported filesystems in the spooling either way